8318217: RISC-V: C2 VectorizedHashCode

ygaevsky opened the pull request:
Hello All,

Please review these changes to support _vectorizedHashCode intrinsic on
RISC-V platform. The patch adds the "scalar" code for the intrinsic without
usage of any RVV instruction but provides manual unrolling of the appropriate
loop. The code with usage of RVV instruction could be added as follow-up of
the patch or independently.

Thanks,
-Yuri Gaevsky

P.S. My OCA has been accepted recently (ygaevsky).

### Correctness checks

Testing: tier1 tests successfully passed on a RISC-V StarFive JH7110 board with Linux.

### Performance results (the numbers for non-ints are similar)

#### StarFive JH7110 board:
```

ArraysHashCode:              without intrinsic      with intrinsic
-------------------------------------------------------------------------------
Benchmark  (size)  Mode  Cnt       Score     Error       Score     Error  Units
-------------------------------------------------------------------------------
multiints       0  avgt   30       2.658 ?   0.001       2.661 ?   0.004  ns/op
multiints       1  avgt   30       4.881 ?   0.011       4.892 ?   0.015  ns/op
multiints       2  avgt   30      16.109 ?   0.041      10.451 ?   0.075  ns/op
multiints       3  avgt   30      14.873 ?   0.068      11.753 ?   0.024  ns/op
multiints       4  avgt   30      17.283 ?   0.078      13.176 ?   0.044  ns/op
multiints       5  avgt   30      19.691 ?   0.136      14.723 ?   0.046  ns/op
multiints       6  avgt   30      21.727 ?   0.166      15.463 ?   0.124  ns/op
multiints       7  avgt   30      23.790 ?   0.126      18.298 ?   0.059  ns/op
multiints       8  avgt   30      23.527 ?   0.116      18.267 ?   0.046  ns/op
multiints       9  avgt   30      27.981 ?   0.303      20.453 ?   0.069  ns/op
multiints      10  avgt   30      26.947 ?   0.215      20.541 ?   0.051  ns/op
multiints      50  avgt   30      95.373 ?   0.588      69.238 ?   0.208  ns/op
multiints     100  avgt   30     177.109 ?   0.525     137.852 ?   0.417  ns/op
multiints     200  avgt   30     341.074 ?   1.363     296.832 ?   0.725  ns/op
multiints     500  avgt   30     847.993 ?   1.713     752.415 ?   1.918  ns/op
multiints    1000  avgt   30    1610.199 ?   5.424    1426.112 ?   3.407  ns/op
multiints   10000  avgt   30   16234.260 ?  26.789   14447.936 ?  26.345  ns/op
multiints  100000  avgt   30  170726.025 ? 184.003  152587.649 ? 381.964  ns/op
-------------------------------------------------------------------------------
```
#### T-Head RVB-ICE board:

```
ArraysHashCode:              without intrinsic      with intrinsic
------------------------------------------------------------------------------
Benchmark  (size)  Mode  Cnt       Score     Error      Score     Error  Units
------------------------------------------------------------------------------
multiints       0  avgt   30       2.780 ?   0.022      2.816 ?   0.038  ns/op
multiints       1  avgt   30       5.073 ?   0.032      5.101 ?   0.064  ns/op
multiints       2  avgt   30      14.656 ?   0.234     10.974 ?   0.118  ns/op
multiints       3  avgt   30      12.890 ?   0.064     14.168 ?   0.096  ns/op
multiints       4  avgt   30      13.715 ?   0.092     12.552 ?   0.188  ns/op
multiints       5  avgt   30      19.068 ?   0.172     13.557 ?   0.164  ns/op
multiints       6  avgt   30      18.863 ?   0.122     14.848 ?   0.086  ns/op
multiints       7  avgt   30      23.155 ?   0.123     17.300 ?   0.360  ns/op
multiints       8  avgt   30      21.656 ?   0.130     19.214 ?   1.689  ns/op
multiints       9  avgt   30      27.223 ?   0.116     20.450 ?   0.088  ns/op
multiints      10  avgt   30      26.194 ?   0.116     19.463 ?   0.411  ns/op
multiints      50  avgt   30      97.904 ?   1.396     69.662 ?   1.729  ns/op
multiints     100  avgt   30     179.607 ?   0.766    137.127 ?   3.156  ns/op
multiints     200  avgt   30     303.845 ?   2.743    235.476 ?   3.446  ns/op
multiints     500  avgt   30     752.295 ?   2.131    571.157 ?   4.258  ns/op
multiints    1000  avgt   30    1404.489 ?   5.839   1048.263 ?   3.473  ns/op
multiints   10000  avgt   30   13797.829 ?  44.821  10344.752 ?  24.183  ns/op
multiints  100000  avgt   30  135067.307 ? 254.361  98806.823 ? 131.562  ns/op
------------------------------------------------------------------------------
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318217](https://bugs.openjdk.org/browse/JDK-8318217): RISC-V: C2 VectorizedHashCode (**Sub-task** - P5)


### Reviewers
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16629/head:pull/16629` \
`$ git checkout pull/16629`

Update a local copy of the PR: \
`$ git checkout pull/16629` \
`$ git pull https://git.openjdk.org/jdk.git pull/16629/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16629`

View PR using the GUI difftool: \
`$ git pr show -t 16629`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16629.diff">https://git.openjdk.org/jdk/pull/16629.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16629#issuecomment-1808090075)

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,107 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  default:                                             break; \
+  } \
+
+  ld(pow31_3_4, ExternalAddress(StubRoutines::riscv::arrays_hashcode_powers_of_31()
```
same comment as the above `ld` of pow31_1_2.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,107 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+                                + 0 * sizeof(jint))); // [31^^3:31^^4]
+
+  bind(WIDE_LOOP);
+  DO_ELEMENT_LOAD(tmp1, 0)
```
Can you add `;` at the end of the statement? similar comments for other DO_ELEMENT_LOAD's

ygaevsky replied:
Agreed.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,107 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+
+  bind(WIDE_LOOP);
+  DO_ELEMENT_LOAD(tmp1, 0)
+  DO_ELEMENT_LOAD(tmp3, 1)
```
Would it help to optimize the perf by moving `DO_ELEMENT_LOAD(tmp3, 1)` after `srli(tmp2, pow31_3_4, 32);`?

ygaevsky replied:
Sure, it makes the code more understandable even though it doesn't improve performance.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,107 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  DO_ELEMENT_LOAD(tmp1, 2)
+  DO_ELEMENT_LOAD(tmp3, 3)
+  srli(tmp2, pow31_1_2, 32);
+  mulw(tmp1, tmp1, tmp2);          // 31^^1 * ary[i+2]
```
Could this line be optimized as `x<<5-x`? Just as TAIL_LOOP below.

ygaevsky replied:
Sure, good catch!

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,107 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  addw(result, result, tmp1);
+  addw(result, result, tmp3);      // 31^^4 * h + 31^^3 * ary[i+0] + 31^^2 * ary[i+1]
+                                   //           + 31^^1 * ary[i+2] + 31^^0 * ary[i+3]
+  subw(chunk, chunk, stride);
```
Could chunk and ary be merged into one variable? so we don't need one sub and one add, but only one add here.

ygaevsky replied:
Could you please carify? I don't understand how that's possible.

Hamlin-Li replied:
chunk is only used to tell if the wide loop is done, which can be done by ary too.

ygaevsky replied:
Oh, thanks: look like I understand now what can be done here and there.

ygaevsky replied:
Done.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,107 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  subw(result, tmp2, result);
+  addw(result, result, tmp1); // result = result + ary[i]
+  subw(cnt, cnt, 1);
+  add(ary, ary, elsize);
```
Similar comment for cnt and ary as chunk and ary above.

ygaevsky replied:
As above, please advice how to do that. IIUC, that's possible with INDEX-REG addressing which is absent in RISC-V. :-(

ygaevsky replied:
Done.

Hamlin-Li commented on src/hotspot/cpu/riscv/riscv.ad:
```diff
@@ -10302,6 +10302,25 @@ instruct array_equalsC(iRegP_R11 ary1, iRegP_R12 ary2, iRegI_R10 result,
 %}
 
+
+instruct arrays_hashcode(iRegP_R11 ary, iRegI_R12 cnt, iRegI_R10 result, immI basic_type,
```
Is it necessary to specify the regs(r11/12/10) here?

ygaevsky replied:
I've just "borrowed" those definitions from other intrinsics around. Do you think we can improve this with iRegP/iRegI?

Hamlin-Li replied:
Seems to me it's not necessary to specify the registers. Can you try it?

ygaevsky replied:
Sure, let me check .

ygaevsky replied:
Done in [this commit](https://github.com/openjdk/jdk/pull/16629/commits/af940acd365677ec3c29a8f066b68b753ad362e4). I've tried the usage of iRegP/iRegI but that caused of the related failure (JVM even didn'r start).

Hamlin-Li replied:
Hey, I saw you already change the code in this patch to not use the specific registers, do you still face the JVM starting issue?

Hamlin-Li replied:
I think the reason might be: with specific register, you can add effect as `USE_KILL ary, USE_KILL cnt`, but without specific register, currently you have to way to do so.
But, in current patch, it does modify the ary and cnt in the intrinsic, so I wonder if the current (lastest) patch is safe enough in all situation.

It maybe be helpful to add 2 new register when matching the instrinsic in ad file, and I guess the register allocator will merge different use of temp register together?
But I still think it's not necessary to specify the register when matching arrays_hashcode in ad file.

ygaevsky replied:
Hi @Hamlin-Li,

My apologies for some delay with an answer.

Without specifying the "concrete" registers for _ary/cnt/result_, as e.g. as follows:
(1) [ iRegP/iRegI ]
```
instruct arrays_hashcode(iRegP ary, iRegI cnt, iRegI result, immI basic_type,
                         iRegLNoSp tmp1, iRegINoSp tmp2,
                         iRegINoSp tmp3, iRegLNoSp tmp4, rFlagsReg cr)
```
or
(2) [ iRegPNoSp/iRegINoSp ]
```
instruct arrays_hashcode(iRegPNoSp ary, iRegINoSp cnt, iRegINoSp result, immI basic_type,
                         iRegLNoSp tmp1, iRegINoSp tmp2,
                         iRegINoSp tmp3, iRegLNoSp tmp4, rFlagsReg cr)
```
it's impossible to use '_USE_KILL_' hint in '_effect_' directive because AD compilation fails:
```
Building target 'images' in configuration 'linux-riscv64-server-release'
/jdk/src/hotspot/cpu/riscv/gc/z/z_riscv.ad(10306) Syntax Error: :In arrays_hashcode only bound registers can be killed: iRegP ary
```
or
```
Building target 'images' in configuration 'linux-riscv64-server-release'
/jdk/src/hotspot/cpu/riscv/gc/z/z_riscv.ad(10306) Syntax Error: :In arrays_hashcode only bound registers can be killed: iRegPNoSp ary
```

IMHO, the usage of '_USE_KILL_' for _ary_/_cnt_ looks reasonable since
we actually do use/modify them both in the assembler body, and avoiding any
hint or usage other hint ('_USE_'?) may be wrong here even if that does
not cause a failure in pre-integration tests. It is interesting that many
intrinsics' definitions in AD files similarly use "concrete" registers in
X86/RISC-V archs, possibly due to the mentioned "_only bound registers can be
killed_" ADLC requirement.

I'd like to mention that applying '_USE_KILL_' to 'result' causes the error
in runtime:
```
---------------  T H R E A D  ---------------

Current thread (0x0000003fc0189130):  JavaThread "C2 CompilerThread0" daemon [_thread_in_native, id=54524, stack(0x0000003f7bc00000,0x0000003f7be00000) (2048K)]


Current CompileTask:
C2:828  117       4       java.lang.String::hashCode (60 bytes)

Stack: [0x0000003f7bc00000,0x0000003f7be00000],  sp=0x0000003f7bdfbba0,  free space=2030k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x5e1ad2]  Node_Backward_Iterator::next()+0x106
V  [libjvm.so+0x5e4040]  PhaseCFG::global_code_motion()+0x242
V  [libjvm.so+0x5e5266]  PhaseCFG::do_global_code_motion()+0x38
V  [libjvm.so+0x455142]  Compile::Code_Gen()+0x198
V  [libjvm.so+0x458168]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0xce4
V  [libjvm.so+0x3a239a]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x278
V  [libjvm.so+0x45d11e]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0x804
V  [libjvm.so+0x45f9e4]  CompileBroker::compiler_thread_loop()+0x4f0
V  [libjvm.so+0x6768b4]  JavaThread::thread_main_inner() [clone .part.0]+0x98
V  [libjvm.so+0xb1d3ea]  Thread::call_run()+0x8c
V  [libjvm.so+0x97253c]  thread_native_entry(Thread*)+0xd0
C  [libc.so.6+0x6a51c]
C  [libc.so.6+0xb7e3e]

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000000028
```
and similarly I do not see in any intrinsics around the usage of '_effect_' hint for '_result_',
again both for X86/RISC-V archs.

I suggest the definition which looks "in sync" with X86 version and similar RISC-V
intrinsics:
```
instruct arrays_hashcode(iRegP_R11 ary, iRegI_R12 cnt, iRegI_R10 result, immI basic_type,
                         iRegLNoSp tmp1, iRegINoSp tmp2,
                         iRegINoSp tmp3, iRegLNoSp tmp4, rFlagsReg cr)
%{
  match(Set result (VectorizedHashCode (Binary ary cnt) (Binary result basic_type)));
  effect(TEMP tmp1, TEMP tmp2, TEMP tmp3, TEMP tmp4,
         USE_KILL ary, USE_KILL cnt, USE basic_type, KILL cr);
```
As an alternative we can use:
```
instruct arrays_hashcode(iRegP ary, iRegI cnt, iRegI_R10 result, immI basic_type,
                         iRegLNoSp tmp1, iRegINoSp tmp2,
                         iRegINoSp tmp3, iRegLNoSp tmp4, rFlagsReg cr)
%{
  match(Set result (VectorizedHashCode (Binary ary cnt) (Binary result basic_type)));
  effect(TEMP tmp1, TEMP tmp2, TEMP tmp3, TEMP tmp4,
        USE basic_type, KILL cr);
```
or as third possibilitiy even introduce additional temp registers and copy
ary/cnt into them at the beginning. But I personally do not like these 2 variants.

Ideally we need to understand why other intrinsics (a) use "concrete" registers and
(b) do not use any hint for 'result' regs, but that look to me as a separate not-so-small
activity. :-(

What do you think? Could C2 regalloc experts give us more details for usage of concrete regs in
intrinsics' definitions?

Hamlin-Li replied:
Thanks for looking into the details.
Based on the information we got till now, yes, we need the concrete registers, otherwise, we need to pass in more tmp register, seems former one is better, let's go with it.

ygaevsky replied:
Thanks @Hamlin-Li, the patch has been updated to use concrete registers.

Hamlin-Li commented on src/hotspot/cpu/riscv/riscv.ad:
```diff
@@ -10302,6 +10302,25 @@ instruct array_equalsC(iRegP_R11 ary1, iRegP_R12 ary2, iRegI_R10 result,
+%{
+  match(Set result (VectorizedHashCode (Binary ary cnt) (Binary result basic_type)));
+  effect(TEMP tmp1, TEMP tmp2, TEMP tmp3, TEMP tmp4, TEMP tmp5, TEMP tmp6,
+         USE_KILL ary, USE_KILL cnt, USE basic_type, KILL cr);
```
should `TEMP_DEF result` be added here?

ygaevsky replied:
Hmm, addition of TEMP_DEF result makes the bencmark results even worse tha without intrinsic (I haven't look at the generated assembler though).

Hamlin-Li replied:
What specific tests were run for this intrinsic implementation to verify the correctness?
BTW, can you add some comments about what java method or bytecode this intrinsic is for?

ygaevsky replied:
> What specific tests were run for this intrinsic implementation to verify the correctness?
- jtreg tier1
- hotspot/jtreg/compiler/intrinsics/TestArraysHashCode.java with -Xcomp

ygaevsky replied:
> BTW, can you add some comments about what java method or bytecode this intrinsic is for?
Done.

Hamlin-Li replied:
```
Hmm, addition of TEMP_DEF result makes the bencmark results even worse tha without intrinsic (I haven't look at the generated assembler though).
```
This seems bit confusing to me.

Hamlin-Li replied:
If we use concrete register for `result`, then there is no need to label it as `TEMP_DEF ` either.

So, in summary, let's go with the concrete registers solution if no one has other opinion.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,107 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  case T_BYTE:    BLOCK_COMMENT("arrays_hashcode(byte) {");          break;
+  case T_SHORT:   BLOCK_COMMENT("arrays_hashcode(short) {");         break;
+  case T_INT:     BLOCK_COMMENT("arrays_hashcode(int) {");           break;
+  default:        BLOCK_COMMENT("arrays_hashcode {");                break;
```
In `C2_MacroAssembler::arrays_hashcode_elsize`, default action is `ShouldNotReachHere();`, should it be consistent here?

ygaevsky replied:
Sure, thanks for catching!

ygaevsky replied:
Done.

Hamlin-Li commented on src/hotspot/cpu/riscv/stubRoutines_riscv.cpp:
```diff
@@ -55,4 +55,13 @@ address StubRoutines::riscv::_string_indexof_linear_ul = nullptr;
 address StubRoutines::riscv::_large_byte_array_inflate = nullptr;
 address StubRoutines::riscv::_method_entry_barrier = nullptr;
 
+ATTRIBUTE_ALIGNED(64) const jint StubRoutines::riscv::_arrays_hashcode_powers_of_31[] =
```
If we use `mv` to replace the `ld` of power values above, these related code could be removed here.

Hamlin-Li commented:
> The code with usage of RVV instruction could be added as follow-up of
the patch or independently.

Hey @ygaevsky, I can work on this real vectorized intrinsic implementation, please let me know how you think about it.
Thanks.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,107 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+
+  beqz(cnt, DONE);
+
+  ld(pow31_1_2, ExternalAddress(StubRoutines::riscv::arrays_hashcode_powers_of_31()
```
Seems not for a  64bit immediate by current cost model of riscv in jdk.
Please ignore this suggestion, and below one.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,108 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+
+  beqz(cnt, DONE);
+
+  lw(pow31_2, ExternalAddress(StubRoutines::riscv::arrays_hashcode_powers_of_31()
```
Now you don't need this `lw` anymore, as 961 will fit in an immediate, so `mv pow31_2, 961` should be fine.

ygaevsky replied:
Agreed.

ygaevsky replied:
Done.

Hamlin-Li commented on src/hotspot/cpu/riscv/stubRoutines_riscv.cpp:
```diff
@@ -55,4 +55,13 @@ address StubRoutines::riscv::_string_indexof_linear_ul = nullptr;
+{
+         923521, // 0x000E1781
+          29791, // 0x0000745F
+            961, // 0x000003C1
```
based on the comment above about `pow31_2 `, line 62-64 can be removed now.

ygaevsky replied:
Done.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,108 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  ld(pow31_3_4, ExternalAddress(StubRoutines::riscv::arrays_hashcode_powers_of_31()
+                                + 0 * sizeof(jint))); // [31^^3:31^^4]
+
+  bind(WIDE_LOOP);
```
Seems in this loop, tmp3 and tmp1 can share one register.

ygaevsky replied:
Done.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,108 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  bind(WIDE_LOOP);
+  mulw(result, result, pow31_3_4); // 31^^4 * h
+  DO_ELEMENT_LOAD(tmp1, 0);
+  srli(tmp2, pow31_3_4, 32);
```
tmp2 can be calculated outside of the loop.

ygaevsky replied:
Done.

RealFYang commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,108 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+                                        Register tmp6,
+                                        BasicType eltype)
+{
+  assert_different_registers(ary, cnt, result, tmp1, tmp2, tmp3, tmp4, tmp5, tmp6);
```
We have two scratch registers t0 (x5) / t1 (x6) which should be considered for use in this assembler function.
These two are reserved from the register allocator and are suitable for keeping some short-lived values in the procedure.

ygaevsky replied:
Done.

ygaevsky commented:
macos-x64/windows-x64 failures look unrelated.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,114 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  case T_BYTE:    BLOCK_COMMENT("arrays_hashcode(byte) {");          break;
+  case T_SHORT:   BLOCK_COMMENT("arrays_hashcode(short) {");         break;
+  case T_INT:     BLOCK_COMMENT("arrays_hashcode(int) {");           break;
+  default:        BLOCK_COMMENT("arrays_hashcode {");                break;
```
Is this `BLOCK_COMMENT("arrays_hashcode {");                break;` necessary?

Hamlin-Li replied:
Yes, as we already have `ShouldNotReachHere();` to guard the default case. How do you think about it?

Hamlin-Li reviewed (commented):
Thanks for updating. Looks good to me.
Just one minor comment.

ygaevsky commented:
@Hamlin-Li, @RealFYang: please take a look at the latest updates when you have time, thanks.

RealFYang commented:
> @Hamlin-Li, @RealFYang: please take a look at the latest updates when you have time, thanks.

Having another look.

RealFYang commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,114 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+
+#define DO_ELEMENT_LOAD(reg, idx) \
+  switch (eltype) { \
+  case T_BOOLEAN: lb(reg, Address(ary, idx * elsize)); break; \
```
Since `T_BOOLEAN` is used to signify unsigned bytes [1], shouldn't we use `lbu` instead of `lb` here? Seems that the existing tests didn't cover this case?

[1] https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/util/ArraysSupport.java#L192

ygaevsky replied:
Agreed.

ygaevsky replied:
Fixed.

RealFYang commented on src/hotspot/cpu/riscv/riscv.ad:
```diff
@@ -10302,6 +10302,24 @@ instruct array_equalsC(iRegP_R11 ary1, iRegP_R12 ary2, iRegI_R10 result,
+// fast ArraysSupport.vectorizedHashCode
+instruct arrays_hashcode(iRegP_R11 ary, iRegI_R12 cnt, iRegI_R10 result, immI basic_type,
+                         iRegLNoSp tmp1, iRegINoSp tmp2,
+                         iRegINoSp tmp3, iRegLNoSp tmp4, rFlagsReg cr)
```
Maybe declare `tmp2` and `tmp3` as `iRegLNoSp`? I see these two are used as 64-bit registers in `C2_MacroAssembler::arrays_hashcode`.

ygaevsky replied:
Agreed.

ygaevsky replied:
Fixed.

RealFYang commented on src/hotspot/cpu/riscv/riscv.ad:
```diff
@@ -10302,6 +10302,24 @@ instruct array_equalsC(iRegP_R11 ary1, iRegP_R12 ary2, iRegI_R10 result,
+%{
+  match(Set result (VectorizedHashCode (Binary ary cnt) (Binary result basic_type)));
+  effect(TEMP tmp1, TEMP tmp2, TEMP tmp3, TEMP tmp4,
+         USE_KILL ary, USE_KILL cnt, USE basic_type, KILL cr);
```
I don't think we need to add `USE basic_type` to the effect list. It is an immediate input which is already there in the match rule.

RealFYang reviewed (changes requested):
Thanks for the update. Several comments remain.

ygaevsky commented:
Thank you for your comments/suggestions, @RealFYang: I've updated the patch except for riscv.ad part where I am not sure we need to do the suggested change for basic_type.

RealFYang commented:
So I tried this on sifive unmatched. Unfortunately, I see some performance regressions with this change.
Before:
```
ArraysHashCode.bytes             1  avgt   15     19.737 ?    5.405  ns/op
ArraysHashCode.bytes            10  avgt   15     56.102 ?    3.191  ns/op
ArraysHashCode.bytes           100  avgt   15    317.126 ?    3.452  ns/op
ArraysHashCode.bytes         10000  avgt   15  28380.470 ?   20.709  ns/op
ArraysHashCode.chars             1  avgt   15     15.532 ?    2.623  ns/op
ArraysHashCode.chars            10  avgt   15     59.603 ?    2.440  ns/op
ArraysHashCode.chars           100  avgt   15    333.995 ?    3.834  ns/op
ArraysHashCode.chars         10000  avgt   15  29464.768 ?   16.751  ns/op
ArraysHashCode.ints              1  avgt   15     16.031 ?    2.820  ns/op
ArraysHashCode.ints             10  avgt   15     59.506 ?    3.980  ns/op
ArraysHashCode.ints            100  avgt   15    335.514 ?    4.695  ns/op
ArraysHashCode.ints          10000  avgt   15  33966.175 ?  929.859  ns/op
ArraysHashCode.multibytes        1  avgt   15      7.840 ?    0.110  ns/op
ArraysHashCode.multibytes       10  avgt   15     34.727 ?    0.547  ns/op
ArraysHashCode.multibytes      100  avgt   15    193.085 ?    0.814  ns/op
ArraysHashCode.multibytes    10000  avgt   15  16610.239 ?   27.290  ns/op
ArraysHashCode.multichars        1  avgt   15      7.853 ?    0.092  ns/op
ArraysHashCode.multichars       10  avgt   15     35.059 ?    0.241  ns/op
ArraysHashCode.multichars      100  avgt   15    203.483 ?    0.413  ns/op
ArraysHashCode.multichars    10000  avgt   15  18819.804 ?   75.487  ns/op
ArraysHashCode.multiints         1  avgt   15      7.878 ?    0.104  ns/op
ArraysHashCode.multiints        10  avgt   15     35.232 ?    0.196  ns/op
ArraysHashCode.multiints       100  avgt   15    211.087 ?    1.914  ns/op
ArraysHashCode.multiints     10000  avgt   15  30172.693 ? 1447.757  ns/op
ArraysHashCode.multishorts       1  avgt   15      7.788 ?    0.046  ns/op
ArraysHashCode.multishorts      10  avgt   15     35.504 ?    0.465  ns/op
ArraysHashCode.multishorts     100  avgt   15    203.530 ?    0.342  ns/op
ArraysHashCode.multishorts   10000  avgt   15  18801.799 ?   77.159  ns/op
ArraysHashCode.shorts            1  avgt   15     19.685 ?    5.413  ns/op
ArraysHashCode.shorts           10  avgt   15     59.583 ?    4.684  ns/op
ArraysHashCode.shorts          100  avgt   15    333.170 ?    5.367  ns/op
ArraysHashCode.shorts        10000  avgt   15  29455.665 ?   13.302  ns/op
```

After:
```
ArraysHashCode.bytes             1  avgt   15     18.575 ?   3.780  ns/op
ArraysHashCode.bytes            10  avgt   15     55.394 ?   4.610  ns/op
ArraysHashCode.bytes           100  avgt   15    340.807 ?   3.387  ns/op
ArraysHashCode.bytes         10000  avgt   15  31506.478 ?  27.694  ns/op
ArraysHashCode.chars             1  avgt   15     15.966 ?   2.291  ns/op
ArraysHashCode.chars            10  avgt   15     56.524 ?   4.301  ns/op
ArraysHashCode.chars           100  avgt   15    343.389 ?   3.272  ns/op
ArraysHashCode.chars         10000  avgt   15  31520.717 ?  13.290  ns/op
ArraysHashCode.ints              1  avgt   15     16.078 ?   3.977  ns/op
ArraysHashCode.ints             10  avgt   15     55.467 ?   2.845  ns/op
ArraysHashCode.ints            100  avgt   15    344.500 ?   3.531  ns/op
ArraysHashCode.ints          10000  avgt   15  36234.542 ?  39.191  ns/op
ArraysHashCode.multibytes        1  avgt   15      7.816 ?   0.072  ns/op
ArraysHashCode.multibytes       10  avgt   15     29.617 ?   0.257  ns/op
ArraysHashCode.multibytes      100  avgt   15    183.986 ?   0.236  ns/op
ArraysHashCode.multibytes    10000  avgt   15  18349.268 ?  28.711  ns/op
ArraysHashCode.multichars        1  avgt   15      7.821 ?   0.050  ns/op
ArraysHashCode.multichars       10  avgt   15     29.293 ?   0.273  ns/op
ArraysHashCode.multichars      100  avgt   15    186.538 ?   0.404  ns/op
ArraysHashCode.multichars    10000  avgt   15  20149.487 ?  87.300  ns/op
ArraysHashCode.multiints         1  avgt   15      7.847 ?   0.044  ns/op
ArraysHashCode.multiints        10  avgt   15     29.765 ?   1.082  ns/op
ArraysHashCode.multiints       100  avgt   15    193.887 ?   0.360  ns/op
ArraysHashCode.multiints     10000  avgt   15  30997.145 ? 420.328  ns/op
ArraysHashCode.multishorts       1  avgt   15      7.856 ?   0.128  ns/op
ArraysHashCode.multishorts      10  avgt   15     29.231 ?   0.434  ns/op
ArraysHashCode.multishorts     100  avgt   15    187.044 ?   0.289  ns/op
ArraysHashCode.multishorts   10000  avgt   15  20146.327 ?  89.985  ns/op
ArraysHashCode.shorts            1  avgt   15     15.162 ?   4.191  ns/op
ArraysHashCode.shorts           10  avgt   15     54.279 ?   2.661  ns/op
ArraysHashCode.shorts          100  avgt   15    343.085 ?   4.204  ns/op
ArraysHashCode.shorts        10000  avgt   15  31536.455 ?  23.874  ns/op
```

ygaevsky commented:
@RealFYang - many thanks for checking my results. I can confirm the regressions found by you.

Sifive Unmatched:
```
Benchmark                   (size)  Mode  Cnt      Score     Error      Score     Error  Units
ArraysHashCode.bytes            10  avgt   15     64.318 ±   2.035     54.062 ±   5.229  ns/op
ArraysHashCode.bytes           100  avgt   15    317.512 ±   1.944    352.153 ±   2.802  ns/op
ArraysHashCode.bytes          1000  avgt   15   2875.723 ±   5.773   3185.938 ±  11.533  ns/op
ArraysHashCode.bytes         10000  avgt   15  28386.639 ±  42.689  31602.956 ± 101.607  ns/op
ArraysHashCode.chars            10  avgt   15     61.976 ±   5.562     52.773 ±   0.572  ns/op
ArraysHashCode.chars           100  avgt   15    332.949 ±   5.650    347.923 ±   2.549  ns/op
ArraysHashCode.chars          1000  avgt   15   2991.818 ±   4.396   3186.808 ±   3.248  ns/op
ArraysHashCode.chars         10000  avgt   15  29528.339 ±  30.613  31580.887 ±  24.868  ns/op
ArraysHashCode.ints             10  avgt   15     62.328 ±   5.592     57.784 ±   5.973  ns/op
ArraysHashCode.ints            100  avgt   15    328.240 ±   0.556    342.637 ±   4.174  ns/op
ArraysHashCode.ints           1000  avgt   15   2984.384 ±   0.865   3175.596 ±   5.490  ns/op
ArraysHashCode.ints          10000  avgt   15  33830.448 ±  55.105  36310.776 ±  39.960  ns/op
ArraysHashCode.multibytes       10  avgt   15     34.593 ±   0.318     29.497 ±   0.164  ns/op
ArraysHashCode.multibytes      100  avgt   15    193.281 ±   0.548    184.273 ±   0.293  ns/op
ArraysHashCode.multibytes     1000  avgt   15   1651.863 ±   7.210   1816.328 ±  15.627  ns/op
ArraysHashCode.multibytes    10000  avgt   15  16653.946 ±  42.054  18371.511 ±  63.512  ns/op
ArraysHashCode.multichars       10  avgt   15     35.104 ±   0.228     29.222 ±   0.165  ns/op
ArraysHashCode.multichars      100  avgt   15    203.873 ±   0.170    187.397 ±   0.249  ns/op
ArraysHashCode.multichars     1000  avgt   15   1900.463 ±   5.472   2018.710 ±   7.485  ns/op
ArraysHashCode.multichars    10000  avgt   15  18991.838 ± 124.180  20257.406 ± 236.705  ns/op
ArraysHashCode.multiints        10  avgt   15     35.232 ±   0.159     29.439 ±   0.199  ns/op
ArraysHashCode.multiints       100  avgt   15    211.201 ±   0.330    197.014 ±   0.242  ns/op
ArraysHashCode.multiints      1000  avgt   15   2225.082 ±  13.261   2363.767 ±   7.716  ns/op
ArraysHashCode.multiints     10000  avgt   15  31441.156 ± 574.230  33215.921 ± 392.809  ns/op
ArraysHashCode.multishorts      10  avgt   15     35.068 ±   0.164     29.237 ±   0.103  ns/op
ArraysHashCode.multishorts     100  avgt   15    203.904 ±   0.393    186.776 ±   0.481  ns/op
ArraysHashCode.multishorts    1000  avgt   15   1897.351 ±   7.325   2019.786 ±   6.929  ns/op
ArraysHashCode.multishorts   10000  avgt   15  18907.910 ±  77.537  20349.999 ± 363.546  ns/op
ArraysHashCode.shorts           10  avgt   15     56.984 ±   0.414     54.080 ±   2.779  ns/op
ArraysHashCode.shorts          100  avgt   15    327.769 ±   2.124    347.356 ±   3.553  ns/op
ArraysHashCode.shorts         1000  avgt   15   2988.008 ±   8.788   3175.180 ±   1.255  ns/op
ArraysHashCode.shorts        10000  avgt   15  29520.151 ±  58.051  31577.025 ±  66.598  ns/op
```

T-Head RVB-ICE:
```
Benchmark                   (size)  Mode  Cnt      Score     Error      Score    Error  Units
ArraysHashCode.bytes            10  avgt   15     51.762 ±   0.378     46.688 ±  0.274  ns/op
ArraysHashCode.bytes           100  avgt   15    282.922 ±   0.891    228.657 ±  1.265  ns/op
ArraysHashCode.bytes          1000  avgt   15   2550.826 ±   3.773   1790.795 ±  3.891  ns/op
ArraysHashCode.bytes         10000  avgt   15  25165.387 ±  72.982  17510.131 ± 30.293  ns/op
ArraysHashCode.chars            10  avgt   15     53.831 ±   0.225     46.425 ±  0.231  ns/op
ArraysHashCode.chars           100  avgt   15    285.555 ±   0.813    237.081 ±  1.216  ns/op
ArraysHashCode.chars          1000  avgt   15   2558.697 ±   6.253   1893.147 ±  3.187  ns/op
ArraysHashCode.chars         10000  avgt   15  25162.198 ±  67.487  18558.521 ± 49.710  ns/op
ArraysHashCode.ints             10  avgt   15     52.046 ±   0.170     46.364 ±  0.240  ns/op
ArraysHashCode.ints            100  avgt   15    285.322 ±   0.924    238.531 ±  0.901  ns/op
ArraysHashCode.ints           1000  avgt   15   2557.016 ±   6.214   1892.898 ±  3.588  ns/op
ArraysHashCode.ints          10000  avgt   15  25272.506 ± 116.785  18599.428 ± 45.966  ns/op
ArraysHashCode.multibytes       10  avgt   15     26.190 ±   0.129     18.942 ±  0.103  ns/op
ArraysHashCode.multibytes      100  avgt   15    160.660 ±   0.400    116.278 ±  0.287  ns/op
ArraysHashCode.multibytes     1000  avgt   15   1366.136 ±   9.128    908.505 ±  2.456  ns/op
ArraysHashCode.multibytes    10000  avgt   15  13290.571 ±  21.784   8975.082 ± 18.118  ns/op
ArraysHashCode.multichars       10  avgt   15     26.630 ±   0.153     19.760 ±  0.178  ns/op
ArraysHashCode.multichars      100  avgt   15    164.371 ±   0.438    118.326 ±  0.334  ns/op
ArraysHashCode.multichars     1000  avgt   15   1399.774 ±   3.369   1031.653 ±  2.678  ns/op
ArraysHashCode.multichars    10000  avgt   15  13321.995 ±  27.186   9653.414 ± 44.167  ns/op
ArraysHashCode.multiints        10  avgt   15     25.878 ±   0.093     19.145 ±  0.217  ns/op
ArraysHashCode.multiints       100  avgt   15    169.129 ±   0.551    126.062 ±  0.239  ns/op
ArraysHashCode.multiints      1000  avgt   15   1405.709 ±   8.575   1046.932 ±  3.110  ns/op
ArraysHashCode.multiints     10000  avgt   15  13712.716 ±  28.671  10414.196 ± 19.044  ns/op
ArraysHashCode.multishorts      10  avgt   15     26.614 ±   0.240     19.742 ±  0.169  ns/op
ArraysHashCode.multishorts     100  avgt   15    164.488 ±   0.296    119.336 ±  0.434  ns/op
ArraysHashCode.multishorts    1000  avgt   15   1396.339 ±   3.062   1032.205 ±  4.604  ns/op
ArraysHashCode.multishorts   10000  avgt   15  13475.962 ±  27.035   9694.721 ± 38.618  ns/op
ArraysHashCode.shorts           10  avgt   15     52.145 ±   0.671     50.525 ±  0.560  ns/op
ArraysHashCode.shorts          100  avgt   15    284.563 ±   0.663    236.216 ±  1.160  ns/op
ArraysHashCode.shorts         1000  avgt   15   2565.527 ±   5.313   1895.385 ±  3.632  ns/op
ArraysHashCode.shorts        10000  avgt   15  25160.261 ±  78.896  18562.613 ± 62.708  ns/op
```

StarFive JH7110
```
Benchmark                   (size)  Mode  Cnt      Score    Error      Score     Error  Units
ArraysHashCode.bytes            10  avgt   15     40.824 ?  0.148     39.565 ?   0.069  ns/op
ArraysHashCode.bytes           100  avgt   15    250.700 ?  1.212    268.188 ?   0.864  ns/op
ArraysHashCode.bytes          1000  avgt   15   2290.265 ?  5.880   2524.649 ?   5.811  ns/op
ArraysHashCode.bytes         10000  avgt   15  22593.288 ?  4.222  25114.007 ?  41.732  ns/op
ArraysHashCode.chars            10  avgt   15     45.736 ?  0.356     40.116 ?   0.043  ns/op
ArraysHashCode.chars           100  avgt   15    261.146 ?  0.383    270.978 ?   0.697  ns/op
ArraysHashCode.chars          1000  avgt   15   2371.921 ?  0.679   2526.244 ?   6.915  ns/op
ArraysHashCode.chars         10000  avgt   15  23421.998 ?  2.523  25095.058 ?  71.641  ns/op
ArraysHashCode.ints             10  avgt   15     45.419 ?  0.007     40.214 ?   0.171  ns/op
ArraysHashCode.ints            100  avgt   15    262.007 ?  1.421    270.151 ?   0.429  ns/op
ArraysHashCode.ints           1000  avgt   15   2371.512 ?  0.402   2525.866 ?   1.160  ns/op
ArraysHashCode.ints          10000  avgt   15  29589.412 ? 13.383  31285.973 ? 163.943  ns/op
ArraysHashCode.multibytes       10  avgt   15     27.079 ?  0.172     22.170 ?   0.114  ns/op
ArraysHashCode.multibytes      100  avgt   15    155.937 ?  0.745    146.639 ?   0.165  ns/op
ArraysHashCode.multibytes     1000  avgt   15   1291.860 ?  2.052   1415.878 ?   2.299  ns/op
ArraysHashCode.multibytes    10000  avgt   15  12777.667 ?  2.627  14123.991 ?  68.454  ns/op
ArraysHashCode.multichars       10  avgt   15     28.005 ?  0.136     22.282 ?   0.077  ns/op
ArraysHashCode.multichars      100  avgt   15    166.277 ?  0.593    151.245 ?   0.524  ns/op
ArraysHashCode.multichars     1000  avgt   15   1432.254 ?  5.252   1513.058 ?   5.612  ns/op
ArraysHashCode.multichars    10000  avgt   15  14125.547 ? 52.368  15030.018 ?  89.649  ns/op
ArraysHashCode.multiints        10  avgt   15     26.678 ?  0.220     22.650 ?   0.260  ns/op
ArraysHashCode.multiints       100  avgt   15    179.278 ?  0.316    165.724 ?   0.606  ns/op
ArraysHashCode.multiints      1000  avgt   15   1605.150 ?  2.885   1684.205 ?   2.334  ns/op
ArraysHashCode.multiints     10000  avgt   15  16517.998 ? 71.869  17220.866 ?  65.161  ns/op
ArraysHashCode.multishorts      10  avgt   15     27.507 ?  0.133     22.270 ?   0.056  ns/op
ArraysHashCode.multishorts     100  avgt   15    166.159 ?  0.377    152.968 ?   0.696  ns/op
ArraysHashCode.multishorts    1000  avgt   15   1430.556 ?  1.569   1509.756 ?   1.608  ns/op
ArraysHashCode.multishorts   10000  avgt   15  14114.634 ?  5.609  14992.519 ?  11.684  ns/op
ArraysHashCode.shorts           10  avgt   15     45.606 ?  0.289     40.080 ?   0.006  ns/op
ArraysHashCode.shorts          100  avgt   15    261.538 ?  0.954    270.748 ?   3.619  ns/op
ArraysHashCode.shorts         1000  avgt   15   2385.091 ?  2.082   2527.360 ?   8.349  ns/op
ArraysHashCode.shorts        10000  avgt   15  23420.647 ?  3.081  25066.912 ?   4.099  ns/op
```

Let me check which updates for the patch caused such results.

ygaevsky commented:
The results of commit 99f91d0 are below.

Sifive Unmatched:
```
Benchmark                   (size)  Mode  Cnt      Score     Error      Score     Error  Units
ArraysHashCode.bytes            10  avgt   15     65.190 ?   0.954     45.527 ?   1.771  ns/op
ArraysHashCode.bytes           100  avgt   15    321.443 ?   5.586    258.807 ?   4.922  ns/op
ArraysHashCode.bytes          1000  avgt   15   2878.206 ?   9.105   2347.219 ?   8.947  ns/op
ArraysHashCode.bytes         10000  avgt   15  28421.840 ?  35.467  23160.425 ?  30.340  ns/op
ArraysHashCode.chars            10  avgt   15     64.544 ?   1.713     50.808 ?   2.629  ns/op
ArraysHashCode.chars           100  avgt   15    338.919 ?   1.623    265.971 ?   4.874  ns/op
ArraysHashCode.chars          1000  avgt   15   2986.972 ?   4.009   2336.699 ?   2.537  ns/op
ArraysHashCode.chars         10000  avgt   15  29474.441 ?  14.634  23161.582 ?  29.067  ns/op
ArraysHashCode.ints             10  avgt   15     57.104 ?   2.517     46.034 ?   0.602  ns/op
ArraysHashCode.ints            100  avgt   15    330.264 ?   4.543    258.327 ?   1.517  ns/op
ArraysHashCode.ints           1000  avgt   15   2995.208 ?   3.188   2339.664 ?   6.849  ns/op
ArraysHashCode.ints          10000  avgt   15  33855.312 ? 115.319  27836.954 ?  27.304  ns/op
ArraysHashCode.multibytes       10  avgt   15     34.378 ?   0.230     27.076 ?   0.108  ns/op
ArraysHashCode.multibytes      100  avgt   15    193.131 ?   0.370    141.907 ?   0.244  ns/op
ArraysHashCode.multibytes     1000  avgt   15   1651.909 ?   7.812   1377.842 ?  10.299  ns/op
ArraysHashCode.multibytes    10000  avgt   15  16620.685 ?  37.854  13960.556 ?  43.473  ns/op
ArraysHashCode.multichars       10  avgt   15     35.104 ?   0.195     26.308 ?   0.127  ns/op
ArraysHashCode.multichars      100  avgt   15    204.391 ?   0.233    144.662 ?   0.337  ns/op
ArraysHashCode.multichars     1000  avgt   15   1902.088 ?   6.922   1579.549 ?   7.266  ns/op
ArraysHashCode.multichars    10000  avgt   15  18905.923 ?  79.263  15952.155 ?  68.664  ns/op
ArraysHashCode.multiints        10  avgt   15     35.111 ?   0.093     26.551 ?   0.264  ns/op
ArraysHashCode.multiints       100  avgt   15    211.251 ?   0.550    153.683 ?   0.208  ns/op
ArraysHashCode.multiints      1000  avgt   15   2223.176 ?   8.982   1927.689 ?   7.075  ns/op
ArraysHashCode.multiints     10000  avgt   15  31567.767 ? 249.609  29463.762 ? 186.245  ns/op
ArraysHashCode.multishorts      10  avgt   15     35.311 ?   0.313     26.372 ?   0.116  ns/op
ArraysHashCode.multishorts     100  avgt   15    203.294 ?   0.241    144.988 ?   0.494  ns/op
ArraysHashCode.multishorts    1000  avgt   15   1898.485 ?   6.704   1579.381 ?   5.600  ns/op
ArraysHashCode.multishorts   10000  avgt   15  18855.850 ?  66.545  15718.005 ?  75.154  ns/op
ArraysHashCode.shorts           10  avgt   15     56.418 ?   0.186     47.488 ?   2.261  ns/op
ArraysHashCode.shorts          100  avgt   15    337.844 ?   1.202    256.671 ?   0.761  ns/op
ArraysHashCode.shorts         1000  avgt   15   2988.457 ?   6.158   2337.570 ?   2.510  ns/op
ArraysHashCode.shorts        10000  avgt   15  29506.107 ?  41.616  23148.772 ?  40.625  ns/op
```

T-Head RVB-ICE:
```
Benchmark                   (size)  Mode  Cnt      Score    Error      Score    Error  Units
ArraysHashCode.bytes            10  avgt   15     53.463 ?  0.274     46.625 ?  0.247  ns/op
ArraysHashCode.bytes           100  avgt   15    280.976 ?  1.478    225.197 ?  1.141  ns/op
ArraysHashCode.bytes          1000  avgt   15   2553.393 ?  4.925   1818.613 ?  3.789  ns/op
ArraysHashCode.bytes         10000  avgt   15  25138.794 ? 39.992  16787.514 ? 59.261  ns/op
ArraysHashCode.chars            10  avgt   15     52.075 ?  0.246     45.924 ?  0.561  ns/op
ArraysHashCode.chars           100  avgt   15    283.441 ?  0.743    237.660 ?  1.074  ns/op
ArraysHashCode.chars          1000  avgt   15   2562.833 ?  3.370   1915.665 ?  4.166  ns/op
ArraysHashCode.chars         10000  avgt   15  25168.219 ? 94.226  18843.917 ? 51.859  ns/op
ArraysHashCode.ints             10  avgt   15     52.126 ?  0.382     46.739 ?  0.366  ns/op
ArraysHashCode.ints            100  avgt   15    283.643 ?  0.901    242.191 ?  0.776  ns/op
ArraysHashCode.ints           1000  avgt   15   2556.508 ?  6.937   1913.271 ?  2.920  ns/op
ArraysHashCode.ints          10000  avgt   15  25171.578 ? 51.725  18835.638 ? 49.785  ns/op
ArraysHashCode.multibytes       10  avgt   15     26.432 ?  0.157     18.762 ?  0.184  ns/op
ArraysHashCode.multibytes      100  avgt   15    160.788 ?  0.484    117.339 ?  0.285  ns/op
ArraysHashCode.multibytes     1000  avgt   15   1366.697 ?  9.217    923.814 ?  4.709  ns/op
ArraysHashCode.multibytes    10000  avgt   15  13360.445 ? 22.830   9350.136 ? 18.251  ns/op
ArraysHashCode.multichars       10  avgt   15     26.732 ?  0.181     19.234 ?  0.136  ns/op
ArraysHashCode.multichars      100  avgt   15    164.043 ?  0.310    117.900 ?  0.386  ns/op
ArraysHashCode.multichars     1000  avgt   15   1398.259 ?  2.765   1030.563 ?  2.701  ns/op
ArraysHashCode.multichars    10000  avgt   15  13331.460 ? 21.356   9749.817 ? 23.566  ns/op
ArraysHashCode.multiints        10  avgt   15     25.972 ?  0.135     18.745 ?  0.155  ns/op
ArraysHashCode.multiints       100  avgt   15    169.487 ?  0.357    125.620 ?  0.330  ns/op
ArraysHashCode.multiints      1000  avgt   15   1399.977 ?  9.000   1036.132 ?  3.237  ns/op
ArraysHashCode.multiints     10000  avgt   15  13760.907 ? 23.137  10324.485 ? 18.437  ns/op
ArraysHashCode.multishorts      10  avgt   15     26.541 ?  0.223     19.389 ?  0.151  ns/op
ArraysHashCode.multishorts     100  avgt   15    163.990 ?  0.301    117.797 ?  0.419  ns/op
ArraysHashCode.multishorts    1000  avgt   15   1402.545 ?  3.285   1031.649 ?  7.023  ns/op
ArraysHashCode.multishorts   10000  avgt   15  13349.611 ? 25.599   9778.011 ? 19.135  ns/op
ArraysHashCode.shorts           10  avgt   15     52.037 ?  0.265     46.881 ?  0.636  ns/op
ArraysHashCode.shorts          100  avgt   15    285.775 ?  0.702    244.200 ?  1.012  ns/op
ArraysHashCode.shorts         1000  avgt   15   2553.894 ?  5.309   1926.098 ?  3.496  ns/op
ArraysHashCode.shorts        10000  avgt   15  25201.063 ? 95.129  18843.485 ? 73.870  ns/op
```

StarFive JH7110
```
Benchmark                   (size)  Mode  Cnt      Score    Error      Score    Error  Units
ArraysHashCode.bytes            10  avgt   15     41.093 ?  0.541     34.051 ?  0.032  ns/op
ArraysHashCode.bytes           100  avgt   15    250.250 ?  0.846    201.460 ?  0.631  ns/op
ArraysHashCode.bytes          1000  avgt   15   2283.792 ?  0.293   1855.048 ?  0.337  ns/op
ArraysHashCode.bytes         10000  avgt   15  22613.649 ? 85.647  18454.512 ? 93.310  ns/op
ArraysHashCode.chars            10  avgt   15     45.441 ?  0.108     34.747 ?  0.008  ns/op
ArraysHashCode.chars           100  avgt   15    261.762 ?  1.081    203.169 ?  0.118  ns/op
ArraysHashCode.chars          1000  avgt   15   2372.976 ?  1.541   1856.964 ?  4.764  ns/op
ArraysHashCode.chars         10000  avgt   15  23429.722 ?  6.530  18390.679 ?  2.956  ns/op
ArraysHashCode.ints             10  avgt   15     45.530 ?  0.284     34.744 ?  0.005  ns/op
ArraysHashCode.ints            100  avgt   15    261.117 ?  0.721    203.332 ?  0.218  ns/op
ArraysHashCode.ints           1000  avgt   15   2373.573 ?  3.175   1856.836 ?  0.223  ns/op
ArraysHashCode.ints          10000  avgt   15  29624.472 ? 44.767  24626.598 ? 54.767  ns/op
ArraysHashCode.multibytes       10  avgt   15     26.975 ?  0.259     19.854 ?  0.114  ns/op
ArraysHashCode.multibytes      100  avgt   15    156.220 ?  0.247    113.744 ?  0.366  ns/op
ArraysHashCode.multibytes     1000  avgt   15   1296.236 ?  7.541   1073.224 ?  4.383  ns/op
ArraysHashCode.multibytes    10000  avgt   15  12779.460 ?  2.007  10593.835 ?  5.531  ns/op
ArraysHashCode.multichars       10  avgt   15     27.520 ?  0.102     19.992 ?  0.054  ns/op
ArraysHashCode.multichars      100  avgt   15    166.026 ?  0.695    117.982 ?  0.639  ns/op
ArraysHashCode.multichars     1000  avgt   15   1430.447 ?  1.517   1165.180 ?  5.783  ns/op
ArraysHashCode.multichars    10000  avgt   15  14134.839 ?  6.270  11499.764 ? 37.546  ns/op
ArraysHashCode.multiints        10  avgt   15     26.872 ?  0.066     20.127 ?  0.083  ns/op
ArraysHashCode.multiints       100  avgt   15    178.919 ?  0.245    132.377 ?  0.484  ns/op
ArraysHashCode.multiints      1000  avgt   15   1607.719 ?  2.903   1339.118 ?  8.704  ns/op
ArraysHashCode.multiints     10000  avgt   15  16390.804 ? 49.820  13706.741 ? 11.994  ns/op
ArraysHashCode.multishorts      10  avgt   15     27.749 ?  0.165     20.011 ?  0.096  ns/op
ArraysHashCode.multishorts     100  avgt   15    166.625 ?  0.592    119.115 ?  0.324  ns/op
ArraysHashCode.multishorts    1000  avgt   15   1429.682 ?  1.607   1165.839 ?  6.013  ns/op
ArraysHashCode.multishorts   10000  avgt   15  14199.682 ?  6.483  11493.880 ?  6.484  ns/op
ArraysHashCode.shorts           10  avgt   15     45.878 ?  0.348     34.768 ?  0.116  ns/op
ArraysHashCode.shorts          100  avgt   15    260.598 ?  0.079    203.937 ?  0.078  ns/op
ArraysHashCode.shorts         1000  avgt   15   2374.712 ?  7.961   1857.542 ?  0.248  ns/op
ArraysHashCode.shorts        10000  avgt   15  23428.899 ?  4.859  18433.195 ? 34.224  ns/op
```
The improvements on SiFive/StarFive came after the move of  all memory loads up to the start of the loop. Differences between Out-of-Order versus In-Order CPUs?

Hamlin-Li commented:
This is interesting, as in another patch (https://github.com/openjdk/jdk/pull/16453), a little perf improvement is acheived in reverse way.

ygaevsky commented:
> This is interesting, as in another patch (#16453), a little perf improvement is acheived in reverse way.

Just for clarity, yesterday at first I added two additional temp registers for memory loads to avoid register dependencies, and that didn't help at all. After that I moved all loads to the start of "wide" loop and exactly that step fixed perfromance regressions on SiFive/StarFive so now the performance numbers on them are consistent with THead ones.

By the the way, could you please provide more details for #16453 perf improvement mentioned above?

RealFYang reviewed (commented):
Hi, glad to see the performance numbers are back to normal. Would you mind two more tweaks? Thanks.

ygaevsky commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,117 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  bne(ary, chunks_end, WIDE_LOOP);
+
+  bind(TAIL);
+  beqz(cnt, DONE);
```
We need this check because after wide "unrolling" loop the cnt could be 0,1,2 or 3, see "don't forget about tail" comment at the line 1502.

RealFYang replied:
The control flow will be directed to `DONE` by the `beqz` check at L1493 when `cnt` is zero.

ygaevsky replied:
I meant other case(s): if '`cnt`' is equal to 4/8/... then we pass the initial `cnt==zero` check at L1493 and go further into wide loop, where the 'cnt' is zero (L1501)- that's why that check is needed.

RealFYang replied:
Yes, I know what you mean. And that's also what I am suggesting in my initial comment: move this `beqz` check immediately after the loop (that is after L1522 in your latest version).

ygaevsky replied:
Ah, got it finally. Nice catch, thanks!

ygaevsky replied:
Fixed.

ygaevsky commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1458,6 +1458,117 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
+  add(chunks_end, ary, chunks_end);
+  andi(cnt, cnt, stride-1); // don't forget about tail!
+
+#define DO_ELEMENT_LOAD(reg, idx) \
```
Good idea, will do, thanks.

ygaevsky replied:
Fixed.

ygaevsky commented:
> Hi, glad to see the performance numbers are back to normal. Would you mind two more tweaks? Thanks.

Thank you for the suggestions, fixed, please take a look.

ygaevsky commented:
The failures above seem unrelated.

RealFYang commented:
Thanks for the update. So I gave it a second try and did some extra tunning. I see some extra 7%+ improvement on licheepi-4a board (T-Head C910) with following small add-on change (no obvious change on unmatched board). This materializes the powers of 31 with direct `mv` instructions and avoids loading elements from `_arrays_hashcode_powers_of_31` array which would involve calculation of the array address.

```
diff --git a/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp b/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp
index 11cbcaa48a1..fe82b7a4e74 100644
--- a/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp
+++ b/src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp
@@ -1493,16 +1493,16 @@ void C2_MacroAssembler::arrays_hashcode(Register ary, Register cnt, Register res

   beqz(cnt, DONE);

-  addiw(pow31_2, zr, 961);       // [31^^2]
   andi(chunks, cnt, ~(stride-1));
   beqz(chunks, TAIL);

+  mv(pow31_4, 923521); // [31^^4]
+  mv(pow31_3, 29791);  // [31^^3]
+  mv(pow31_2, 961);    // [31^^2]
+
   slli(chunks_end, chunks, chunks_end_shift);
   add(chunks_end, ary, chunks_end);
   andi(cnt, cnt, stride-1);      // don't forget about tail!
-  ld(pow31_4, ExternalAddress(StubRoutines::riscv::arrays_hashcode_powers_of_31()
-                              + 0 * sizeof(jint))); // [31^^3:31^^4]
-  srli(pow31_3, pow31_4, 32);

   bind(WIDE_LOOP);
   mulw(result, result, pow31_4); // 31^^4 * h
```

1. Without addon fix:
```
Benchmark                   (size)  Mode  Cnt      Score    Error  Units
ArraysHashCode.bytes             1  avgt   15     21.327 ?  0.035  ns/op
ArraysHashCode.bytes            10  avgt   15     33.195 ?  0.166  ns/op
ArraysHashCode.bytes           100  avgt   15    154.175 ?  3.433  ns/op
ArraysHashCode.bytes         10000  avgt   15  12318.680 ? 25.131  ns/op
ArraysHashCode.chars             1  avgt   15     20.965 ?  0.598  ns/op
ArraysHashCode.chars            10  avgt   15     33.097 ?  0.117  ns/op
ArraysHashCode.chars           100  avgt   15    153.510 ?  0.280  ns/op
ArraysHashCode.chars         10000  avgt   15  11881.690 ? 44.507  ns/op
ArraysHashCode.ints              1  avgt   15     21.330 ?  0.070  ns/op
ArraysHashCode.ints             10  avgt   15     33.409 ?  0.225  ns/op
ArraysHashCode.ints            100  avgt   15    154.254 ?  0.650  ns/op
ArraysHashCode.ints          10000  avgt   15  11833.894 ? 73.945  ns/op
ArraysHashCode.multibytes        1  avgt   15      3.468 ?  0.046  ns/op
ArraysHashCode.multibytes       10  avgt   15     12.412 ?  0.126  ns/op
ArraysHashCode.multibytes      100  avgt   15     75.963 ?  0.267  ns/op
ArraysHashCode.multibytes    10000  avgt   15   6587.068 ? 53.064  ns/op
ArraysHashCode.multichars        1  avgt   15      3.437 ?  0.042  ns/op
ArraysHashCode.multichars       10  avgt   15     13.019 ?  0.118  ns/op
ArraysHashCode.multichars      100  avgt   15     82.657 ?  0.244  ns/op
ArraysHashCode.multichars    10000  avgt   15   6743.844 ? 80.474  ns/op
ArraysHashCode.multiints         1  avgt   15      3.409 ?  0.036  ns/op
ArraysHashCode.multiints        10  avgt   15     13.102 ?  0.140  ns/op
ArraysHashCode.multiints       100  avgt   15     82.864 ?  1.002  ns/op
ArraysHashCode.multiints     10000  avgt   15   7107.843 ? 69.506  ns/op
ArraysHashCode.multishorts       1  avgt   15      3.475 ?  0.033  ns/op
ArraysHashCode.multishorts      10  avgt   15     12.923 ?  0.108  ns/op
ArraysHashCode.multishorts     100  avgt   15     82.498 ?  0.450  ns/op
ArraysHashCode.multishorts   10000  avgt   15   6744.477 ? 22.576  ns/op
ArraysHashCode.shorts            1  avgt   15     21.337 ?  0.077  ns/op
ArraysHashCode.shorts           10  avgt   15     33.236 ?  0.114  ns/op
ArraysHashCode.shorts          100  avgt   15    154.099 ?  0.421  ns/op
ArraysHashCode.shorts        10000  avgt   15  11876.918 ? 41.767  ns/op
```

2. With add-on change:
```
Benchmark                   (size)  Mode  Cnt      Score    Error  Units
ArraysHashCode.bytes             1  avgt   15     21.311 ?  0.036  ns/op
ArraysHashCode.bytes            10  avgt   15     32.113 ?  0.124  ns/op
ArraysHashCode.bytes           100  avgt   15    150.476 ?  0.635  ns/op
ArraysHashCode.bytes         10000  avgt   15  11639.521 ? 16.383  ns/op
ArraysHashCode.chars             1  avgt   15     21.329 ?  0.041  ns/op
ArraysHashCode.chars            10  avgt   15     32.315 ?  0.466  ns/op
ArraysHashCode.chars           100  avgt   15    151.996 ?  1.008  ns/op
ArraysHashCode.chars         10000  avgt   15  10957.449 ? 23.898  ns/op
ArraysHashCode.ints              1  avgt   15     21.323 ?  0.035  ns/op
ArraysHashCode.ints             10  avgt   15     32.416 ?  0.170  ns/op
ArraysHashCode.ints            100  avgt   15    152.277 ?  0.555  ns/op
ArraysHashCode.ints          10000  avgt   15  11019.286 ? 53.589  ns/op
ArraysHashCode.multibytes        1  avgt   15      3.450 ?  0.026  ns/op
ArraysHashCode.multibytes       10  avgt   15     12.204 ?  0.171  ns/op
ArraysHashCode.multibytes      100  avgt   15     78.433 ?  0.357  ns/op
ArraysHashCode.multibytes    10000  avgt   15   6654.488 ? 19.664  ns/op
ArraysHashCode.multichars        1  avgt   15      3.443 ?  0.043  ns/op
ArraysHashCode.multichars       10  avgt   15     12.364 ?  0.087  ns/op
ArraysHashCode.multichars      100  avgt   15     78.246 ?  0.540  ns/op
ArraysHashCode.multichars    10000  avgt   15   6455.363 ? 30.115  ns/op
ArraysHashCode.multiints         1  avgt   15      3.441 ?  0.019  ns/op
ArraysHashCode.multiints        10  avgt   15     12.493 ?  0.063  ns/op
ArraysHashCode.multiints       100  avgt   15     78.485 ?  0.587  ns/op
ArraysHashCode.multiints     10000  avgt   15   6843.608 ? 82.197  ns/op
ArraysHashCode.multishorts       1  avgt   15      3.466 ?  0.029  ns/op
ArraysHashCode.multishorts      10  avgt   15     12.369 ?  0.144  ns/op
ArraysHashCode.multishorts     100  avgt   15     78.172 ?  0.580  ns/op
ArraysHashCode.multishorts   10000  avgt   15   6446.791 ? 13.104  ns/op
ArraysHashCode.shorts            1  avgt   15     20.971 ?  0.574  ns/op
ArraysHashCode.shorts           10  avgt   15     32.002 ?  0.642  ns/op
ArraysHashCode.shorts          100  avgt   15    152.359 ?  0.692  ns/op
ArraysHashCode.shorts        10000  avgt   15  10968.816 ? 31.404  ns/op
```

ygaevsky commented:
Avoiding memory accesses is always good idea, so your suggestion makes perfect sense, thanks, fixed. I've re-checked that the performance is at least not worse on THead/SiFive/StarFive.

RealFYang reviewed (approved):
Updated change looks good to me. Thanks for your patience.

ygaevsky commented:
> Updated change looks good to me. Thanks for your patience.

Thank you for the thorough review and suggestions, @RealFYang.

Hamlin-Li reviewed (approved):
LGTM. Thanks!

ygaevsky commented:
Thanks for your review/suggestions, @Hamlin-Li!

ygaevsky commented:
/sponsor

ygaevsky commented:
/integrate

VladimirKempik commented:
/sponsor
